File: experiments/mathexpr-ref-go/mathexpr.go
// Package mathexpr provides a math expression parser and evaluator.
//
// It implements a pipeline: Tokenize -> Parse -> Evaluate, composed
// into a single Calc() function as the public API.
package mathexpr

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Token types
// ---------------------------------------------------------------------------

// Token represents a lexical token with a Kind and a Value.
type Token struct {
	Kind  string
	Value string
}

// NewToken creates a Token with the given kind and value.
func NewToken(kind, value string) Token {
	return Token{Kind: kind, Value: value}
}

// ---------------------------------------------------------------------------
// AST types
// ---------------------------------------------------------------------------

// AstNode is the interface implemented by all AST node types.
type AstNode interface {
	nodeType() string
}

// NumberLiteral represents a numeric value (leaf node).
type NumberLiteral struct {
	Val float64
}

func (n NumberLiteral) nodeType() string { return "number" }

// UnaryExpr represents a unary operator applied to an operand.
type UnaryExpr struct {
	Op      string
	Operand AstNode
}

func (u UnaryExpr) nodeType() string { return "unary" }

// BinaryExpr represents a binary operator applied to left and right operands.
type BinaryExpr struct {
	Op    string
	Left  AstNode
	Right AstNode
}

func (b BinaryExpr) nodeType() string { return "binary" }

// NewNumberLiteral creates a NumberLiteral node.
func NewNumberLiteral(value float64) NumberLiteral {
	return NumberLiteral{Val: value}
}

// NewUnaryExpr creates a UnaryExpr node.
func NewUnaryExpr(op string, operand AstNode) UnaryExpr {
	return UnaryExpr{Op: op, Operand: operand}
}

// NewBinaryExpr creates a BinaryExpr node.
func NewBinaryExpr(op string, left, right AstNode) BinaryExpr {
	return BinaryExpr{Op: op, Left: left, Right: right}
}

// ---------------------------------------------------------------------------
// Tokenizer
// ---------------------------------------------------------------------------

// Tokenize converts a math expression string into a sequence of tokens.
func Tokenize(input string) ([]Token, error) {
	tokens := []Token{}
	i := 0
	runes := []rune(input)

	for i < len(runes) {
		ch := runes[i]

		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		if ch == '(' {
			tokens = append(tokens, NewToken("lparen", "("))
			i++
			continue
		}

		if ch == ')' {
			tokens = append(tokens, NewToken("rparen", ")"))
			i++
			continue
		}

		if ch == '+' {
			tokens = append(tokens, NewToken("plus", "+"))
			i++
			continue
		}

		if ch == '-' {
			tokens = append(tokens, NewToken("minus", "-"))
			i++
			continue
		}

		if ch == '*' {
			if i+1 < len(runes) && runes[i+1] == '*' {
				tokens = append(tokens, NewToken("power", "**"))
				i += 2
			} else {
				tokens = append(tokens, NewToken("star", "*"))
				i++
			}
			continue
		}

		if ch == '/' {
			tokens = append(tokens, NewToken("slash", "/"))
			i++
			continue
		}

		if ch == '%' {
			tokens = append(tokens, NewToken("percent", "%"))
			i++
			continue
		}

		if isDigit(ch) || ch == '.' {
			num := ""
			hasDot := false
			for i < len(runes) && (isDigit(runes[i]) || runes[i] == '.') {
				if runes[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("Unexpected character '.' at position %d", i)
					}
					hasDot = true
				}
				num += string(runes[i])
				i++
			}
			tokens = append(tokens, NewToken("number", num))
			continue
		}

		return nil, fmt.Errorf("Unexpected character '%c' at position %d", ch, i)
	}

	return tokens, nil
}

func isDigit(ch rune) bool {
	return ch >= '0' && ch <= '9'
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

// Parse converts a sequence of tokens into an AST using recursive descent.
func Parse(tokens []Token) (AstNode, error) {
	pos := 0

	peek := func() *Token {
		if pos < len(tokens) {
			t := tokens[pos]
			return &t
		}
		return nil
	}

	advance := func() Token {
		t := tokens[pos]
		pos++
		return t
	}

	expect := func(kind string) (Token, error) {
		t := peek()
		if t == nil || t.Kind != kind {
			got := "end of input"
			if t != nil {
				got = t.Kind
			}
			return Token{}, fmt.Errorf("Expected %s but got %s", kind, got)
		}
		return advance(), nil
	}

	// Forward declarations for mutual recursion
	var parseAddSub func() (AstNode, error)
	var parseMulDiv func() (AstNode, error)
	var parsePower func() (AstNode, error)
	var parseUnary func() (AstNode, error)
	var parseAtom func() (AstNode, error)

	// Level 1: addition and subtraction (lowest precedence)
	parseAddSub = func() (AstNode, error) {
		left, err := parseMulDiv()
		if err != nil {
			return nil, err
		}
		for {
			t := peek()
			if t == nil || (t.Kind != "plus" && t.Kind != "minus") {
				break
			}
			opToken := advance()
			op := "+"
			if opToken.Kind == "minus" {
				op = "-"
			}
			right, err := parseMulDiv()
			if err != nil {
				return nil, err
			}
			left = NewBinaryExpr(op, left, right)
		}
		return left, nil
	}

	// Level 2: multiplication, division, modulo
	parseMulDiv = func() (AstNode, error) {
		left, err := parsePower()
		if err != nil {
			return nil, err
		}
		for {
			t := peek()
			if t == nil || (t.Kind != "star" && t.Kind != "slash" && t.Kind != "percent") {
				break
			}
			opToken := advance()
			var op string
			switch opToken.Kind {
			case "star":
				op = "*"
			case "slash":
				op = "/"
			default:
				op = "%"
			}
			right, err := parsePower()
			if err != nil {
				return nil, err
			}
			left = NewBinaryExpr(op, left, right)
		}
		return left, nil
	}

	// Level 3: exponentiation (right-associative)
	parsePower = func() (AstNode, error) {
		base, err := parseUnary()
		if err != nil {
			return nil, err
		}
		t := peek()
		if t != nil && t.Kind == "power" {
			advance()
			exponent, err := parsePower() // right-recursive for right-associativity
			if err != nil {
				return nil, err
			}
			return NewBinaryExpr("**", base, exponent), nil
		}
		return base, nil
	}

	// Level 4: unary minus
	parseUnary = func() (AstNode, error) {
		t := peek()
		if t != nil && t.Kind == "minus" {
			advance()
			operand, err := parseUnary() // allow chained unary: --x
			if err != nil {
				return nil, err
			}
			return NewUnaryExpr("-", operand), nil
		}
		return parseAtom()
	}

	// Level 5: atoms -- numbers and parenthesized expressions
	parseAtom = func() (AstNode, error) {
		t := peek()

		if t == nil {
			return nil, fmt.Errorf("Unexpected end of input")
		}

		if t.Kind == "number" {
			advance()
			val, err := strconv.ParseFloat(t.Value, 64)
			if err != nil {
				return nil, fmt.Errorf("invalid number: %s", t.Value)
			}
			return NewNumberLiteral(val), nil
		}

		if t.Kind == "lparen" {
			advance()
			expr, err := parseAddSub()
			if err != nil {
				return nil, err
			}
			_, err = expect("rparen")
			if err != nil {
				return nil, err
			}
			return expr, nil
		}

		return nil, fmt.Errorf("Unexpected token: %s '%s'", t.Kind, t.Value)
	}

	ast, err := parseAddSub()
	if err != nil {
		return nil, err
	}

	if pos < len(tokens) {
		remaining := tokens[pos]
		return nil, fmt.Errorf("Unexpected token after expression: %s '%s'", remaining.Kind, remaining.Value)
	}

	return ast, nil
}

// ---------------------------------------------------------------------------
// Evaluator
// ---------------------------------------------------------------------------

// Evaluate evaluates an AST node to produce a numeric result.
func Evaluate(node AstNode) (float64, error) {
	switch n := node.(type) {
	case NumberLiteral:
		return n.Val, nil
	case UnaryExpr:
		operand, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -operand, nil
	case BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}
		switch n.Op {
		case "+":
			return left + right, nil
		case "-":
			return left - right, nil
		case "*":
			return left * right, nil
		case "**":
			return math.Pow(left, right), nil
		case "/":
			if right == 0 {
				return 0, fmt.Errorf("Division by zero")
			}
			return left / right, nil
		case "%":
			if right == 0 {
				return 0, fmt.Errorf("Modulo by zero")
			}
			return math.Mod(left, right), nil
		}
	}
	return 0, fmt.Errorf("unknown node type")
}

// ---------------------------------------------------------------------------
// Calc -- public API
// ---------------------------------------------------------------------------

// Calc evaluates a math expression string and returns the result.
// It composes Tokenize -> Parse -> Evaluate.
func Calc(expression string) (float64, error) {
	if strings.TrimSpace(expression) == "" {
		return 0, fmt.Errorf("Empty expression")
	}
	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}
	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}
	return Evaluate(ast)
}

File: research/experiments/mathexpr-ref-python/mathexpr.py
"""
Math expression parser and evaluator.

Pipeline: tokenize -> parse -> evaluate

Translated from the Type-O reference implementation (TypeScript).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional


# ---------------------------------------------------------------------------
# token-types
# ---------------------------------------------------------------------------

@dataclass
class Token:
    kind: str  # "number"|"plus"|"minus"|"star"|"slash"|"percent"|"power"|"lparen"|"rparen"
    value: str


def token(kind: str, value: str) -> Token:
    return Token(kind=kind, value=value)


# ---------------------------------------------------------------------------
# ast-types
# ---------------------------------------------------------------------------

@dataclass
class NumberLiteral:
    type: str  # always "number"
    value: float


@dataclass
class UnaryExpr:
    type: str  # always "unary"
    op: str
    operand: "AstNode"


@dataclass
class BinaryExpr:
    type: str  # always "binary"
    op: str
    left: "AstNode"
    right: "AstNode"


AstNode = NumberLiteral | UnaryExpr | BinaryExpr


def number_literal(value: float) -> NumberLiteral:
    return NumberLiteral(type="number", value=value)


def unary_expr(op: str, operand: AstNode) -> UnaryExpr:
    return UnaryExpr(type="unary", op=op, operand=operand)


def binary_expr(op: str, left: AstNode, right: AstNode) -> BinaryExpr:
    return BinaryExpr(type="binary", op=op, left=left, right=right)


# ---------------------------------------------------------------------------
# tokenizer
# ---------------------------------------------------------------------------

def _is_digit(ch: str) -> bool:
    return "0" <= ch <= "9"


def tokenize(input_str: str) -> List[Token]:
    tokens: List[Token] = []
    i = 0

    while i < len(input_str):
        ch = input_str[i]

        if ch in (" ", "\t", "\n", "\r"):
            i += 1
            continue

        if ch == "(":
            tokens.append(token("lparen", "("))
            i += 1
            continue

        if ch == ")":
            tokens.append(token("rparen", ")"))
            i += 1
            continue

        if ch == "+":
            tokens.append(token("plus", "+"))
            i += 1
            continue

        if ch == "-":
            tokens.append(token("minus", "-"))
            i += 1
            continue

        if ch == "*":
            if i + 1 < len(input_str) and input_str[i + 1] == "*":
                tokens.append(token("power", "**"))
                i += 2
            else:
                tokens.append(token("star", "*"))
                i += 1
            continue

        if ch == "/":
            tokens.append(token("slash", "/"))
            i += 1
            continue

        if ch == "%":
            tokens.append(token("percent", "%"))
            i += 1
            continue

        if _is_digit(ch) or ch == ".":
            num = ""
            has_dot = False
            while i < len(input_str) and (_is_digit(input_str[i]) or input_str[i] == "."):
                if input_str[i] == ".":
                    if has_dot:
                        raise ValueError(f"Unexpected character '.' at position {i}")
                    has_dot = True
                num += input_str[i]
                i += 1
            tokens.append(token("number", num))
            continue

        raise ValueError(f"Unexpected character '{ch}' at position {i}")

    return tokens


# ---------------------------------------------------------------------------
# parser
# ---------------------------------------------------------------------------

def parse(tokens: List[Token]) -> AstNode:
    pos = 0

    def peek() -> Optional[Token]:
        nonlocal pos
        if pos < len(tokens):
            return tokens[pos]
        return None

    def advance() -> Token:
        nonlocal pos
        t = tokens[pos]
        pos += 1
        return t

    def expect(kind: str) -> Token:
        t = peek()
        if t is None or t.kind != kind:
            got = t.kind if t else "end of input"
            raise ValueError(f"Expected {kind} but got {got}")
        return advance()

    # Level 1: addition and subtraction (lowest precedence)
    def parse_add_sub() -> AstNode:
        left = parse_mul_div()
        while True:
            t = peek()
            if t is None:
                break
            if t.kind == "plus":
                advance()
                right = parse_mul_div()
                left = binary_expr("+", left, right)
            elif t.kind == "minus":
                advance()
                right = parse_mul_div()
                left = binary_expr("-", left, right)
            else:
                break
        return left

    # Level 2: multiplication, division, modulo
    def parse_mul_div() -> AstNode:
        left = parse_power()
        while True:
            t = peek()
            if t is None:
                break
            if t.kind == "star":
                advance()
                right = parse_power()
                left = binary_expr("*", left, right)
            elif t.kind == "slash":
                advance()
                right = parse_power()
                left = binary_expr("/", left, right)
            elif t.kind == "percent":
                advance()
                right = parse_power()
                left = binary_expr("%", left, right)
            else:
                break
        return left

    # Level 3: exponentiation (right-associative)
    def parse_power() -> AstNode:
        base = parse_unary()
        t = peek()
        if t is not None and t.kind == "power":
            advance()
            exponent = parse_power()  # right-recursive for right-associativity
            return binary_expr("**", base, exponent)
        return base

    # Level 4: unary minus
    def parse_unary() -> AstNode:
        t = peek()
        if t is not None and t.kind == "minus":
            advance()
            operand = parse_unary()  # allow chained unary: --x
            return unary_expr("-", operand)
        return parse_atom()

    # Level 5: atoms - numbers and parenthesized expressions
    def parse_atom() -> AstNode:
        t = peek()

        if t is None:
            raise ValueError("Unexpected end of input")

        if t.kind == "number":
            advance()
            return number_literal(float(t.value))

        if t.kind == "lparen":
            advance()
            expr = parse_add_sub()
            expect("rparen")
            return expr

        raise ValueError(f"Unexpected token: {t.kind} '{t.value}'")

    ast = parse_add_sub()

    if pos < len(tokens):
        remaining = tokens[pos]
        raise ValueError(
            f"Unexpected token after expression: {remaining.kind} '{remaining.value}'"
        )

    return ast


# ---------------------------------------------------------------------------
# evaluator
# ---------------------------------------------------------------------------

def evaluate(node: AstNode) -> float:
    if isinstance(node, NumberLiteral):
        return node.value

    if isinstance(node, UnaryExpr):
        return -evaluate(node.operand)

    # BinaryExpr
    left = evaluate(node.left)
    right = evaluate(node.right)

    if node.op == "+":
        return left + right
    if node.op == "-":
        return left - right
    if node.op == "*":
        return left * right
    if node.op == "**":
        return left ** right
    if node.op == "/":
        if right == 0:
            raise ValueError("Division by zero")
        return left / right
    # node.op == "%"
    if right == 0:
        raise ValueError("Modulo by zero")
    return left % right


# ---------------------------------------------------------------------------
# calc (public API)
# ---------------------------------------------------------------------------

def calc(expression: str) -> float:
    """Evaluate a math expression string and return the numeric result."""
    if expression.strip() == "":
        raise ValueError("Empty expression")
    tokens = tokenize(expression)
    ast = parse(tokens)
    return evaluate(ast)
